File: scripts/split-solve-batch-file.py
#!/usr/bin/env python3
"""Split concatenated TilePz JSON batch into N chunk files."""

from __future__ import annotations

import json
import sys
from pathlib import Path


def split_json_docs(text: str) -> list[str]:
    out: list[str] = []
    depth = 0
    start = -1
    in_str = False
    esc = False
    for i, ch in enumerate(text):
        if in_str:
            if esc:
                esc = False
            elif ch == "\\":
                esc = True
            elif ch == '"':
                in_str = False
        else:
            if ch == '"':
                in_str = True
            elif ch == "{":
                if depth == 0:
                    start = i
                depth += 1
            elif ch == "}":
                depth -= 1
                if depth == 0 and start >= 0:
                    out.append(text[start : i + 1])
                    start = -1
    return out


def main() -> None:
    if len(sys.argv) < 2:
        print("Usage: python scripts/split-solve-batch-file.py <batch.txt> [chunks=4]", file=sys.stderr)
        sys.exit(1)
    src = Path(sys.argv[1])
    chunks = int(sys.argv[2]) if len(sys.argv) > 2 else 4
    docs = split_json_docs(src.read_text(encoding="utf-8"))
    per = (len(docs) + chunks - 1) // chunks
    base = str(src).replace(".txt", "")
    written = []
    for c in range(chunks):
        slice_docs = docs[c * per : (c + 1) * per]
        if not slice_docs:
            break
        out = Path(f"{base} - batch{c + 1}.txt")
        out.write_text("\n\n".join(slice_docs) + "\n", encoding="utf-8")
        first = json.loads(slice_docs[0]).get("levelId", "?")
        last = json.loads(slice_docs[-1]).get("levelId", "?")
        written.append((out, len(slice_docs), first, last))
    print(json.dumps({"total": len(docs), "chunks": written}, indent=2, default=str))


if __name__ == "__main__":
    main()

File: scripts/split-solve-batch-file.js
#!/usr/bin/env node
/**
 * Split a concatenated TilePz solve export into N chunk files (JSON docs preserved in order).
 *
 * Usage:
 *   node scripts/split-solve-batch-file.js "data/tilepz solves 5x6 13 jun 2026.txt" --chunks 4
 *   node scripts/split-solve-batch-file.js input.txt --chunks 4 --out "data/tilepz solves 5x6 13 jun 2026 - batch"
 */

'use strict';

const fs = require('fs');
const path = require('path');

function splitJsonDocs(s) {
  const out = [];
  let depth = 0;
  let start = -1;
  let inStr = false;
  let esc = false;
  for (let i = 0; i < s.length; i++) {
    const ch = s[i];
    if (inStr) {
      if (esc) esc = false;
      else if (ch === '\\') esc = true;
      else if (ch === '"') inStr = false;
    } else {
      if (ch === '"') inStr = true;
      else if (ch === '{') {
        if (depth === 0) start = i;
        depth++;
      } else if (ch === '}') {
        depth--;
        if (depth === 0 && start >= 0) {
          out.push(s.slice(start, i + 1));
          start = -1;
        }
      }
    }
  }
  return out;
}

function usage() {
  console.error('Usage: node scripts/split-solve-batch-file.js <batch.txt> [--chunks N] [--out prefix]');
  process.exit(1);
}

function main() {
  const args = process.argv.slice(2);
  if (!args.length) usage();

  const input = args[0];
  let chunks = 4;
  let outPrefix = null;

  for (let i = 1; i < args.length; i++) {
    if (args[i] === '--chunks' && args[i + 1]) {
      chunks = Math.max(1, Number(args[i + 1]) || 4);
      i++;
    } else if (args[i] === '--out' && args[i + 1]) {
      outPrefix = args[i + 1];
      i++;
    }
  }

  const abs = path.isAbsolute(input) ? input : path.join(process.cwd(), input);
  if (!fs.existsSync(abs)) {
    console.error('File not found:', abs);
    process.exit(1);
  }

  const docs = splitJsonDocs(fs.readFileSync(abs, 'utf8'));
  if (!docs.length) {
    console.error('No JSON documents found in', input);
    process.exit(1);
  }

  const base = outPrefix || abs.replace(/\.txt$/i, '');
  const per = Math.ceil(docs.length / chunks);
  const written = [];

  for (let c = 0; c < chunks; c++) {
    const slice = docs.slice(c * per, (c + 1) * per);
    if (!slice.length) break;
    const outPath = `${base} - batch${c + 1}.txt`;
    fs.writeFileSync(outPath, `${slice.join('\n\n')}\n`, 'utf8');
    const first = JSON.parse(slice[0]).levelId || '?';
    const last = JSON.parse(slice[slice.length - 1]).levelId || '?';
    written.push({ path: outPath, count: slice.length, first, last });
  }

  console.log(JSON.stringify({
    source: input,
    totalDocs: docs.length,
    chunks: written.length,
    files: written.map((w) => ({
      file: path.relative(process.cwd(), w.path).replace(/\\/g, '/'),
      count: w.count,
      first: w.first,
      last: w.last,
    })),
  }, null, 2));
}

main();
